packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -1063,38 +1035,44 @@ function renderRoot(
         // Don't bother with a very short suspense time.
         if (msUntilTimeout > 10) {
           if (workInProgressRootHasPendingPing) {
-            // This render was pinged but we didn't get to restart earlier so try
-            // restarting now instead.
-            prepareFreshStack(root, expirationTime);
-            return renderRoot.bind(null, root, expirationTime);
+            const lastPingedTime = root.lastPingedTime;
+            if (lastPingedTime === NoWork || lastPingedTime >= expirationTime) {
+              // This render was pinged but we didn't get to restart earlier so
+              // try restarting now instead.
+              root.lastPingedTime = expirationTime;
+              prepareFreshStack(root, expirationTime);
+              return;
+            }
           }
 
-          const nextKnownPendingLevel = root.nextKnownPendingLevel;
-          if (nextKnownPendingLevel !== NoWork) {
-            // There's lower priority work. It might be unsuspended. Try rendering
-            // at that level.
-            return renderRoot.bind(null, root, nextKnownPendingLevel);
+          const nextTime = getNextRootExpirationTimeToWorkOn(root);
+          if (nextTime !== NoWork && nextTime !== expirationTime) {
+            // There's additional work on this root.
+            return;
```

Comment:
It doesn't really hurt to schedule a callback that is just going to get canceled, right? Maybe it's ok to let it schedule a callback just in case we don't get called for a while to other work, we might as well commit?

Reply:
https://github.com/facebook/react/pull/16715/files/6ace553825626774d9d9325be84389fe7fe5e33a..fd253c9b0662c48f9b780c6d48bc884c99cace7f#r322551325